pkg/controller/poolcoordinator/poolcoordinator_controller.go

```diff
@@ -0,0 +1,378 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	informercoordv1 "k8s.io/client-go/informers/coordination/v1"
+	v1 "k8s.io/client-go/informers/core/v1"
+	client "k8s.io/client-go/kubernetes"
+	leaselisterv1 "k8s.io/client-go/listers/coordination/v1"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/controller/poolcoordinator/constant"
+	"github.com/openyurtio/openyurt/pkg/controller/poolcoordinator/utils"
+	"github.com/openyurtio/openyurt/pkg/controller/poolcoordinator/webhook"
+)
+
+type Controller struct {
+	client        client.Interface
+	nodeInformer  v1.NodeInformer
+	nodeSynced    cache.InformerSynced
+	nodeLister    listerv1.NodeLister
+	leaseInformer informercoordv1.LeaseInformer
+	leaseSynced   cache.InformerSynced
+	leaseLister   leaselisterv1.LeaseNamespaceLister
+	nodepoolMap   *utils.NodepoolMap
+
+	nodeUpdateQueue     workqueue.Interface
+	nodePoolUpdateQueue workqueue.Interface
+}
+
+type nodeTaintItem struct {
+	nodeName string
+	taint    bool
+}
+
+type nodePoolItem struct {
+	nodepoolName string
+	nodeName     string
+	add          bool
+}
+
+type LeaseDelegatedCounter struct {
+	v    map[string]int
+	lock sync.Mutex
+}
+
+var (
+	ctl *Controller
+
+	ldc *LeaseDelegatedCounter
+)
+
+func NewLeaseDelegatedCounter() *LeaseDelegatedCounter {
+	return &LeaseDelegatedCounter{
+		v: make(map[string]int),
+	}
+}
+
+func (dc *LeaseDelegatedCounter) Inc(name string) {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	if dc.v[name] >= constant.LeaseDelegationThreshold {
+		return
+	}
+	dc.v[name] += 1
+}
+
+func (dc *LeaseDelegatedCounter) Dec(name string) {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	if dc.v[name] > 0 {
+		dc.v[name] -= 1
+	}
+}
+
+func (dc *LeaseDelegatedCounter) Reset(name string) {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	dc.v[name] = 0
+}
+
+func (dc *LeaseDelegatedCounter) Counter(name string) int {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	return dc.v[name]
+}
+
+func onLeaseCreate(n interface{}) {
+	nl := n.(*coordv1.Lease)
+	if nl.Namespace != corev1.NamespaceNodeLease {
+		return
+	}
+	//klog.Infof("new lease: %v\n", nl)
+
+	if val, ok := nl.Annotations[constant.DelegateHeartBeat]; ok {
+		if val == "true" {
+			ldc.Inc(nl.Name)
+		}
+	}
+}
+
+func onLeaseUpdate(o interface{}, n interface{}) {
+	//ol := o.(*coordv1.Lease)
+	nl := n.(*coordv1.Lease)
+	if nl.Namespace != corev1.NamespaceNodeLease {
+		return
+	}
+
+	//klog.Infof("updated lease for: %v\n", nl.Name)
+
+	//oval, ook := ol.Annotations[constant.DelegateHeartBeat]
+	nval, nok := nl.Annotations[constant.DelegateHeartBeat]
+
+	item := nodeTaintItem{
+		nodeName: nl.Name,
+		taint:    false,
+	}
+	if nok && nval == "true" {
+		ldc.Inc(nl.Name)
+		if ldc.Counter(nl.Name) >= constant.LeaseDelegationThreshold {
+			item.taint = true
+			GetController().nodeUpdateQueue.Add(item)
```

Comment:
in order to improve the readability of code, maybe we can add node name instead nodeTaintItem  into nodeUpdateQueue.